### PR TITLE
feat: eventbridge api-d cloudevents

### PR DESCRIPTION
**Pull request by embano1**
Closes: #2057

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by bfreiberg** [comment]
Hi @embano1 is this PR ready for review? It is marked as a draft and I just want to make sure.

**Comment by embano1** [comment]
Thx @bfreiberg ! Will do tomorrow, this was mainly for @boyney123 to take a look since he has related work here.

**Comment by embano1** [comment]
**TODO:**

- [x] code
- [x] pattern json
- [ ] README

**Comment by bfreiberg** [comment]
Hi @embano1 , did you have a chance to work on this?

**Comment by embano1** [comment]
@bfreiberg sorry for delaying this one. I'd love to have @boyney123 confirm the content of the pattern since he's writing a blog post. If this code is good, I'll amend with the example JSON, so please hold for a bit more :)

File: eventbridge-api-destinations-cloudevents-cdk-typescript/lib/eventbridge-api-destinations-cloudevents-cdk-typescript-stack.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      {
+        connection: conn,
+        endpoint: webhookUrl,
+        rateLimitPerSecond: 1,
+      }
+    )
+
+    // structured mode CloudEvent
+    const structuredCeTransformer = {
+      specversion: "1.0",
```
Comment (boyney123) [inline eventbridge-api-destinations-cloudevents-cdk-typescript/lib/eventbridge-api-destinations-cloudevents-cdk-typescript-stack.ts:36]: @embano1 should we update this to 1.0.2 is that the latest one?

**Comment by boyney123** [comment]
nice pattern @embano1 thanks, will include link on my blog post, guess this can't go out without custom header support?

**Comment by embano1** [comment]
Yes, this needs /hold for now. 

`specversion: 1.0` is fine because

```
specversion
Type: String

Description: The version of the CloudEvents specification which the event uses. This enables the interpretation of the context. Compliant event producers MUST use a value of 1.0 when referring to this version of the specification.

Currently, this attribute will only have the 'major' and 'minor' version numbers included in it. This allows for 'patch' changes to the specification to be made without changing this property's value in the serialization. Note: for 'release candidate' releases a suffix might be used for testing purposes.
```

**Comment by embano1** [comment]
@boyney123 @bfreiberg finalized the code. Please kindly review.

File: eventbridge-api-destinations-cloudevents-cdk-typescript/example-pattern.json [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    "level": "200",
+    "framework": "CDK",
+    "introBox": {
+        "headline": "How it works",
+        "text": [
+            "This pattern demonstrates using EventBridge API destinations to convert EventBridge events to CloudEvents using binary or structured mode encoding in CloudEvents."
+        ]
+    },
+    "gitHub": {
+        "template": {
```
Comment (bfreiberg) [inline eventbridge-api-destinations-cloudevents-cdk-typescript/example-pattern.json:14]: The `templateFile` is missing.

File: eventbridge-api-destinations-cloudevents-cdk-typescript/lib/eventbridge-api-destinations-cloudevents-cdk-typescript-stack.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    })
+
+    // binary mode CloudEvent
+    const binaryModeCeTarget = new targets.ApiDestination(apiDestination, {
+      deadLetterQueue: dlq,
+      retryAttempts: 3,
+      event: RuleTargetInput.fromEventPath('$.detail'),
+      headerParameters: {
+        "Content-Type": "application/json; charset=UTF-8", // also EventBridge default
+        "ce-specversion": "1.0",
```
Comment (bfreiberg) [inline eventbridge-api-destinations-cloudevents-cdk-typescript/lib/eventbridge-api-destinations-cloudevents-cdk-typescript-stack.ts:62]: see comment above regarding current version

File: eventbridge-api-destinations-cloudevents-cdk-typescript/test/eventbridge-api-destinations-cloudevents-cdk-typescript.test.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+// import * as cdk from 'aws-cdk-lib';
+// import { Template } from 'aws-cdk-lib/assertions';
+// import * as EventbridgeApiDestinationsCloudeventsCdkTypescript from '../lib/eventbridge-api-destinations-cloudevents-cdk-typescript-stack';
+
+// example test. To run these tests, uncomment this file along with the
+// example resource in lib/eventbridge-api-destinations-cloudevents-cdk-typescript-stack.ts
+test('SQS Queue Created', () => {
```
Comment (bfreiberg) [inline eventbridge-api-destinations-cloudevents-cdk-typescript/test/eventbridge-api-destinations-cloudevents-cdk-typescript.test.ts:7]: Was this supposed to be uncommented?

File: eventbridge-api-destinations-cloudevents-cdk-typescript/lib/eventbridge-api-destinations-cloudevents-cdk-typescript-stack.ts (hunk 3 above)
Comment (embano1) [reply eventbridge-api-destinations-cloudevents-cdk-typescript/lib/eventbridge-api-destinations-cloudevents-cdk-typescript-stack.ts:62]: `1.0` is correct (as per my comment above)

File: eventbridge-api-destinations-cloudevents-cdk-typescript/test/eventbridge-api-destinations-cloudevents-cdk-typescript.test.ts (hunk 4 above)
Comment (embano1) [reply eventbridge-api-destinations-cloudevents-cdk-typescript/test/eventbridge-api-destinations-cloudevents-cdk-typescript.test.ts:7]: bootstrap files :) will remove

File: eventbridge-api-destinations-cloudevents-cdk-typescript/example-pattern.json (hunk 2 above)
Comment (embano1) [reply eventbridge-api-destinations-cloudevents-cdk-typescript/example-pattern.json:14]: Let me know if the below is correct

File: eventbridge-api-destinations-cloudevents-cdk-typescript/example-pattern.json [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            "repoURL": "https://github.com/aws-samples/serverless-patterns/tree/main/eventbridge-api-destinations-cloudevents-cdk-typescript",
+            "templateURL": "serverless-patterns/eventbridge-api-destinations-cloudevents-cdk-typescript",
+            "projectFolder": "eventbridge-api-destinations-cloudevents-cdk-typescript",
+            "templateFile": "lib/eventbridge-api-destinations-cloudevents-cdk-typescript-stack.ts"
```
Comment (bfreiberg) [inline eventbridge-api-destinations-cloudevents-cdk-typescript/example-pattern.json:18]: There is an issue when the templateFile contains the projectFolder name. Could you please rename the stack file to something that doesn't contain the folder name.

**Comment by embano1** [comment]
@bfreiberg done

**Comment by bfreiberg** [comment]
Great, thank you for your contribution. The pattern will be merged by a Developer Advocate

**Comment by embano1** [comment]
Thx a ton @bfreiberg, great review! cc/ @boyney123 read to merge since this was released https://aws.amazon.com/about-aws/whats-new/2024/02/amazon-eventbridge-api-destinations-content-type-header-customization/

**Pull request merged**